testing.md

```diff
@@ -0,0 +1,299 @@
+
+- A _unit test_ tests a single function, method or class. External dependencies
+  of the test subject are generally mocked out using, for example,
+  [`package:mockito`](https://github.com/fibulwinter/dart-mockito/blob/master/README.md).
+  Unit tests generally do not read from/write to disk, render to screen and do
+  not receive user actions from outside the process running the test. The goal
+  of a unit-test is to verify the correctness of a piece of logic under a
+  variety of conditions.
+- A _widget test_ (in other UI frameworks referred to as _component test_) tests
+  a single widget. Testing a widget involves multiple classes and requires a
+  test environment that makes the widget feel like it's part of a real UI. For
+  example, it should be able to receive and respond to user actions and events,
+  perform layout, and instantiate child widgets. A widget test is
+  therefore more comprehensive than a unit test. However, like a unit test, a
+  widget test's environment is replaced with an implementation much simpler than
+  a full-blown UI system. The goal of a widget test is to verify that it looks
+  and interacts as expected.
+- An _end-to-end test_ tests a complete application or a large part of an
+  application. Generally an _end-to-end test_ runs on a real device or an OS
+  emulator. The test code does not run in the same process (or even the same
+  device) as the test subject. The goal of an end-to-end test is to verify that
+  the application functions correctly as a whole, that all the widgets it is
+  composed of integrate with each other as expected (which is why another term
+  for these kinds of tests is _integration test_). Another important aspect
+  end-to-end tests verify is application performance.
+
+Here is a table summarizing the tradeoffs concerning the choice between
+different kinds of tests:
+
+|                      | Unit   | Widget | End-to-end |
+|----------------------|--------|--------|------------|
+| **Effectiveness**    | Low    | Higher | Highest    |
+| **Maintenance cost** | Low    | Higher | Highest    |
+| **Dependencies**     | Few    | More   | Lots       |
+| **Execution speed**  | Quick  | Slower | Slowest    |
+
+As a rule of thumb a well-tested application has a very high number of unit
+and widget tests, tracked by line coverage, and a good number of end-to-end
+tests covering all the important usage scenarios.
+
+
+## Unit testing
+
+**NOTE: `flutter test` currently only works on Linux**
+
+To learn how to write unit tests in Dart refer to
+[`package:test` documentation](https://github.com/dart-lang/test/blob/master/README.md).
+To run a test that depends on the Flutter framework use the `flutter test`
+command instead of `pub run test`.
+
+Example:
+
+Add this file to `test/unit_test.dart`:
+
+```dart
+import 'package:test/test.dart';
+
+void main() {
+  test('this is a unit test', () {
+    expect(42, 42);
+  });
+}
+```
+
+Run `flutter test test/unit_test.dart`.
```

Comment:
But the above test doesn't depend on flutter framework. Couldn't I have just run "dart test/unit_test.dart" ?